onnxruntime/test/providers/qnn/logical_comp_ops_test.cc

```diff
@@ -0,0 +1,207 @@
+    constexpr float qdq_scale = 0.0038f;
+
+    auto* input0 = builder.MakeInput<float>(shape, -1.0f, 1.0f);
+    auto* input1 = builder.MakeInput<float>(shape, -1.0f, 1.0f);
+    auto* output = builder.MakeOutput();
+
+    // input -> Q -> DQ -> Op
+    auto* qdq0_output = AddQDQNodePair<InputQType>(builder, input0, qdq_scale, zero_point);
+    auto* qdq1_output = AddQDQNodePair<InputQType>(builder, input1, qdq_scale, zero_point);
+
+    // Op -> output
+    builder.AddNode(op_type, {qdq0_output, qdq1_output}, {output});
+  };
+}
+
+// Runs a model with a logical operator on the QNN CPU backend. Checks the graph node assignment, and that inference
+// outputs for QNN EP and CPU EP match.
+static void RunCPULogicalOpTest(const std::string& op_type, const std::vector<int64_t>& shape,
+                                ExpectedEPNodeAssignment expected_ep_assignment, const char* test_description,
+                                int opset = 17) {
+  ProviderOptions provider_options;
+
+#if defined(_WIN32)
+  provider_options["backend_path"] = "QnnCpu.dll";
+#else
+  provider_options["backend_path"] = "libQnnCpu.so";
+#endif
+
+  constexpr int expected_nodes_in_partition = 1;
+  RunQnnModelTest(BuildLogicalOpTestCase(op_type, shape),
+                  provider_options,
+                  opset,
+                  expected_ep_assignment,
+                  expected_nodes_in_partition,
+                  test_description);
+}
+
+// Runs a model with a logical operator on the QNN HTP backend. Checks the graph node assignment, and that inference
+// outputs for QNN EP and CPU EP match.
+template <typename QuantType>
+static void RunQDQLogicalOpTest(const std::string& op_type, const std::vector<int64_t>& shape,
+                                ExpectedEPNodeAssignment expected_ep_assignment, const char* test_description,
+                                int opset = 17) {
+  ProviderOptions provider_options;
+#if defined(_WIN32)
+  provider_options["backend_path"] = "QnnHtp.dll";
+#else
+  provider_options["backend_path"] = "libQnnHtp.so";
+#endif
+
+  constexpr int expected_nodes_in_partition = 1;
+  RunQnnModelTest(BuildQDQLogicalOpTestCase<QuantType>(op_type, shape),
+                  provider_options,
+                  opset,
+                  expected_ep_assignment,
+                  expected_nodes_in_partition,
+                  test_description);
+}
+
+//
+// CPU tests:
+//
+
+TEST_F(QnnCPUBackendTests, LogicalOpEqual4D) {
+  RunCPULogicalOpTest("Equal", {1, 3, 16, 16}, ExpectedEPNodeAssignment::All, "LogicalOpEqual4D");
+}
+
+TEST_F(QnnCPUBackendTests, LogicalOpGreater4D) {
+  RunCPULogicalOpTest("Greater", {1, 3, 16, 16}, ExpectedEPNodeAssignment::All, "LogicalOpGreater4D");
+}
+
+// TODO: Add support for GreaterOrEqual.
+// According to the ONNX spec, logical operators that are implemented as
+// functions (e.g., LessOrEqual or GreaterOrEqual) do not have output type/shape inference.
+// We need to handle this uniquely.
```

Comment:
To make it clear. The op is supported if the output shape is set. This is covered by existing UTs for CPU, e.g MathOpTest.GreaterOrEqual_broadcastAB.
It's not supported here because the output shape is not set.